concretizer: option to disable compiler mixing between runtime deps

scheibelp opened the pull request:
You can set 

```
concretizer:
  compiler_mixing: false
```

then e.g.:

```
$ spack spec chai%c=oneapi ^umpire%c=gcc
==> Error: failed to concretize `chai %c=oneapi ^umpire %c=gcc` for the following reasons:
     1. Compiler mixing is disabled
```

(even with this setting on, mixing is allowed when a dependency is an external, and only enforced within transitive link run/tree, i.e. the root unification set)

TODOs

- [x] `virtual_on_edge` is not derived (or does not need to always be derived) for reused packages (so this is only guaranteed if reuse is disabled)
  - (Update September 3) this is addressed (and tested) by https://github.com/spack/spack/pull/51135/commits/3a8ce4a2101b22973c50411186ab9c18c806babc
- [ ] (it was agreed to punt this) See https://github.com/spack/spack/issues/51244 and https://github.com/spack/spack/pull/51135#issue-3307026940: this needs a way to allow users to set e.g. a Fortran compiler on the root even when the root doesn't depend on fortran, and then propagate that choice to dependencies that do depend on it
    * Right now with this PR, if you set the fortran compiler on any dependency that needs Fortran, then all of them will be consistent (that is also true on account of the concretizer mandating 1 fortran compiler maximum, but this PR gives those semantics to C/C++ as well).

scheibelp committed:
Basic attempt to disable c-compiler mixing between link deps (when both are not external)

Signed-off-by: Peter Scheibel <scheibel1@llnl.gov>

scheibelp committed:
control mixing for fortran/cxx

Signed-off-by: Peter Scheibel <scheibel1@llnl.gov>

scheibelp committed:
add in config option and send it to asp

Signed-off-by: Peter Scheibel <scheibel1@llnl.gov>

haampie commented:
This is quite elegant :+1: 

Should also work in the future with LLVM-based compilers using `libstdc++` as a runtime library because of how `gcc-runtime` is modeled: it only makes a copy of compiler runtime libraries and has no `c`, `cxx` or `fortran` dependency.

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3029,6 +3029,12 @@ def setup(
+        check_cfg = spack.config.get("concretizer:compiler_mixing")
+        if not isinstance(check_cfg, bool):
+            check_cfg = True
+        should_mix = 1 if check_cfg else 0
```
```suggestion
        should_mix = 1 if spack.config.get("concretizer:compiler_mixing", True) else 0
```

haampie commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1566,6 +1566,29 @@ compiler_used_as_a_library(node(X, Child), Hash) :-
     CompilerNode1 != CompilerNode2.
 
+error(100, "Compiler mixing is disabled") :-
+    attr("virtual_on_edge", PN1, CompilerNode1, "c"),
+    attr("virtual_on_edge", PN2, CompilerNode2, "c"),
```
Bit of inconsistent naming compared to the rest of the file.

`ParentNode` + `ChildNode` seems more common.

haampie commented on lib/spack/spack/schema/concretizer.py:
```diff
@@ -61,6 +61,7 @@
             "unify": {
                 "oneOf": [{"type": "boolean"}, {"type": "string", "enum": ["when_possible"]}]
             },
+            "compiler_mixing": {"type": "boolean"},
```
Can you add `"description": "..."`? It's not common right now, but I'd like to add it since it shows up in editors during autocomplete and when hovering `compiler_mixing: ...`.

scheibelp committed:
review edits

Signed-off-by: Peter Scheibel <scheibel1@llnl.gov>

scheibelp commented:
>> Should also work in the future with LLVM-based compilers .
>> .. there's no c, cxx or fortran dependency.

> I'm not clear how this PR would work with that change: there would still be virtual_on_edge...c/cxx/fortran even if there's no c/cxx/fortran dep? If you have a link to this PR (if it's in progress) or discussion I could read it

I think I understand better now:

* Most packages will have c/cxx/fortran dependencies
* `llvm` will be injecting `gcc-runtime` in the future
* While `gcc-runtime` will depend on `gcc`, it won't be a language dependency, so the check added here wont trip on `X->gcc-runtime`

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1566,6 +1566,29 @@ compiler_used_as_a_library(node(X, Child), Hash) :-
+    attr("depends_on", ParentNode, ChildNode, "link"),
+    not external(ChildNode),
+    CompilerNode1 != CompilerNode2,
+    compiler_mixing(0).
```
This can probably be condensed to:
```suggestion
error(100, "Compiler mixing is disabled") :-
    attr("virtual_on_edge", ParentNode, CompilerNode1, Language),
    attr("virtual_on_edge", ChildNode, CompilerNode2, Language),
    attr("depends_on", ParentNode, ChildNode, "link"),
    language(Language),
    not external(ChildNode),
    CompilerNode1 != CompilerNode2,
    compiler_mixing(0).
```

alalazo replied:
I'll check if I can think of an equivalent formulation that is faster.

alalazo reviewed (commented):
I benchmarked this against:

* **Spack:** 1.1.0.dev0 (https://github.com/spack/spack/commit/c46eebdb46e9b058b08e6ecee72048fede777c93)
* **Builtin repo:** https://github.com/spack/spack-packages/commit/e3d72e6ae4ad19a9fe8fc82488c43a701e739d95
* **Python:** 3.13.2
* **Platform:** linux-ubuntu20.04-icelake

[radiuss.disable_mixing_on.csv](https://github.com/user-attachments/files/21741233/radiuss.disable_mixing_on.csv)
[radiuss.disable_mixing_no_disabling.csv](https://github.com/user-attachments/files/21741234/radiuss.disable_mixing_no_disabling.csv)
[radiuss.develop.csv](https://github.com/user-attachments/files/21741235/radiuss.develop.csv)

When mixing is disabled the same spec can be 30% slower to concretize, otherwise there is no difference. For the record I had a configuration with 4 gcc and 2 llvm compilers

<img width="16000" height="3200" alt="radiuss" src="https://github.com/user-attachments/assets/fe2eaeea-bf1a-4247-a4d0-439e7c384aaa" />

Grounding doesn't seem much of an issue, as it is solving in this case.

scheibelp committed:
consolidate asp clauses; add unit test

scheibelp committed:
allow individual packages to override compiler mixing enforcement

scheibelp committed:
auto style format

scheibelp committed:
partial model

scheibelp committed:
typo fixed; model now sane

scheibelp committed:
simplify varnames

scheibelp committed:
allow_mixing nodes need to sever both above and below

scheibelp committed:
control mixing across siblings

scheibelp committed:
update comment

scheibelp committed:
redo constraints in terms of unification sets

scheibelp committed:
add comment

scheibelp committed:
intermediate work testing toolchains for applying sibling deps

scheibelp committed:
another attempt at sibling compiler control that fails

scheibelp committed:
Merge branch 'develop' into disable-compiler-mixing2

scheibelp committed:
rm sibling mixing test

scheibelp committed:
add statements to account for reused nodes and their compiler

scheibelp committed:
apply config options to reuse

scheibelp committed:
add (failing) test

scheibelp committed:
intermediate work - I cannot get the concretizer to derive a node() for the compiler when it is on a reused node

scheibelp committed:
since I cant figure out how to infer node id (if its possible) infer compiler name/version as an identifier

scheibelp committed:
some cleanup

scheibelp committed:
auto style

becker33 commented on lib/spack/spack/schema/packages.py:
```diff
@@ -168,6 +168,7 @@
                 # attribute, it could be useful to set it here
                 "package_attributes": package_attributes,
                 "variants": variants,
+                "allow_compiler_mixing": {"type": "boolean", "default": False},
```
Is this intended to support packages that fail to build with certain compilers? I think that might not be sufficiently flexible for things that can build with some compilers (and should have mixing enforced for those compilers) but not others.

I wonder if this case would be better supported by a "strong preference" mode for `concretizer:compiler_mixing` instead of an exception to the rule.

scheibelp replied:
> Is this intended to support packages that fail to build with certain compilers?

yes

> I think that might not be sufficiently flexible for things that can build with some compilers (and should have mixing enforced for those compilers) but not others.

I think for example you are saying if a package does not build with `%oneapi`, then we should use `%gcc`/`%clang` if parents are using `%oneapi`, but we should not use `%gcc` if parents are using `%clang`.

I figured when I wrote this that if some kinds of mixing were ok (or necessary) for a package, that other kinds would be acceptable too.

> I wonder if this case would be better supported by a "strong preference" mode for `concretizer:compiler_mixing` instead of an exception to the rule.

I'm not quite sure what that means in terms of implementation: I can think of a couple ways to encourage/force the concretizer to match `parent%clang` to `child%clang` with the implementation here, but did you have a change in mind?

becker33 replied:
I think the simplest implementation would be to have a high priority optimization criteria for minimizing the count. You would compute it the same way as here, but using a `#min` optimization statement.

It would probably require something slightly fancy to ensure we treat 0 and 1 as equally good, so that the solver doesn't turn off options to avoid using a compiler at all.

scheibelp replied:
I think we already have at least one weighting criteria encouraging fewer compilers:

```
% Minimize the number of compilers used on nodes
compiler_penalty(PackageNode, C-1) :-
   C = #count { CompilerNode : node_compiler(PackageNode, CompilerNode) },
   node_compiler(PackageNode, _).
```

(although I don't think this counts compilers on reused nodes, but I don't think that was what you were most-concerned about)

Minimizing number of compilers tends to be fairly "soft" in terms of discouraging mixing: once any node starts using a different compiler, the concretizer stops caring at all about mixing. I think an edge mismatch minimization would be more robust.

alalazo replied:
This configuration seems confusing to me:
```yaml
concretizer:
    compiler_mixing: false
packages:
    hdf5:
        allow_compiler_mixing: true
```
for the following reasons:
1. The configuration in `concretizer` seems to be contradicted by the one in `packages`
2. We need to inspect 2 sections to understand what is the behavior with compiler mixing (I would keep everything within the `concretizer` section)
3. It is not clear what would be the behavior with respect to dependencies and dependents of `hdf5`. Do they need to use the same compiler as `hdf5`? How is the compiler for `hdf5` selected?

becker33 replied:
What about something like:

```
concretizer:
  compiler_mixing:
    allow: false
    exceptions:
    - hdf5
```
This has the benefit of keeping it all together in one place and is fairly legible. The downside is the `exceptions` block is a no-op when `allow: true`, not an exception anymore. So it's not the cleanest.

And then we should document that this means `hdf5` and `hdf5` only (not its dependencies) can be built with a different compiler from the rest of the DAG forest.

scheibelp replied:
I'm happy to go with above proposal or, to address

> The downside is the exceptions block is a no-op when allow: true

the schema could be

```
concretizer:
  # if this is true, everything can mix
  # if this is false, nothing can mix
  # if this is a list of strings, only the packages mentioned in the list can be mixed
  compiler_mixing: [bool, list] 
```

scheibelp replied:
At the moment I've implemented https://github.com/spack/spack/pull/51135#discussion_r2379993355

Right now if you have

```
concretizer:
  compiler_mixing: []
```

Then it would be universally disabled (i.e. the same as `compiler_mixing: False`) so potentially ambiguous. In any case consolidating all the config into the `concretizer:` section is easy and I'm happy to tweak it however folks like.

becker33 replied:
I like the current syntax better well enough, and definitely better than the previous iteration. Checked with some users who agreed ("but mostly I'm just glad it's going in")

scheibelp committed:
count name/version and constraint per-language, not name/version/language

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1566,6 +1566,35 @@ compiler_used_as_a_library(node(X, Child), Hash) :-
+% language if mixing is disabled
+error(100, "Compiler mixing is disabled") :-
+    unification_set_compiler("root", _, _, Z),
+    #count { X,Y : unification_set_compiler("root", X, Y, Z) } > 1.
```
```suggestion
error(100, "Compiler mixing is disabled") :-
    unification_set_compiler("root", _, _, Language),
    #count { CompilerName,CompilerVersion : unification_set_compiler("root", CompilerName, CompilerVersion, Language) } > 1.
```
I think using the full variable names as used in the generating rule will make this much more legible.

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3098,6 +3102,13 @@ def setup(
         self.provider_requirements()
         self.external_packages()
 
+        for pkg_name, d in spack.config.get("packages").items():
+            if pkg_name == "all":
+                continue
```
We should add something that makes clear to users that setting this under `all` isn't valid.

scheibelp replied:
The packages schema omits this setting for `all`, but I thought it was worth making that explicit here. I've added a test and a comment for that.

becker33 commented on lib/spack/spack/test/concretization/core.py:
```diff
@@ -468,6 +468,33 @@ def test_mixing_compilers_only_affects_subdag(self):
             assert x.satisfies("%clang") is not expected_gcc
             assert x.satisfies("%gcc") is expected_gcc
 
+    def test_disable_mixing0(self):
+        with spack.config.override("concretizer", {"compiler_mixing": False}):
+            with pytest.raises(spack.error.UnsatisfiableSpecError):
+                spack.concretize.concretize_one("dt-diamond%clang ^dt-diamond-bottom%gcc")
```
```suggestion
    def test_disable_mixing_prevents_mixing(self):
        with spack.config.override("concretizer", {"compiler_mixing": False}):
            with pytest.raises(spack.error.UnsatisfiableSpecError):
                spack.concretize.concretize_one("dt-diamond%clang ^dt-diamond-bottom%gcc")
```

becker33 commented on lib/spack/spack/test/concretization/core.py:
```diff
@@ -468,6 +468,33 @@ def test_mixing_compilers_only_affects_subdag(self):
+            with pytest.raises(spack.error.UnsatisfiableSpecError):
+                spack.concretize.concretize_one("dt-diamond%clang ^dt-diamond-bottom%gcc")
+
+    def test_disable_mixing1(self):
```
```suggestion
    def test_disable_mixing_override_by_package(self):
```

becker33 commented on lib/spack/spack/test/concretization/core.py:
```diff
@@ -468,6 +468,33 @@ def test_mixing_compilers_only_affects_subdag(self):
+                assert root["dt-diamond-bottom"].satisfies("%gcc")
+                assert root["dt-diamond-left"].satisfies("%clang")
+
+    def test_disable_mixing2(self, mutable_database):
```
```suggestion
    def test_disable_mixing_reuse(self, mutable_database):
```

becker33 reviewed (changes requested):
This seems to be helping a lot for my use cases.

It does seem to be allowing overrides from package preferences in ways that I wouldn't expect, I'll see if I can create a reproducer. But the net result is we get the concretizations that we want currently.

scheibelp committed:
Merge branch 'develop' into disable-compiler-mixing2

scheibelp committed:
better test names

scheibelp committed:
better var names in asp

scheibelp committed:
add test related to config option on the 'all' package

scheibelp committed:
auto-style fix

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3065,6 +3065,10 @@ def setup(
 
+        should_mix = 1 if spack.config.get("concretizer:compiler_mixing", True) else 0
+        for lang in ["c", "cxx", "fortran"]:
+            self.gen.fact(fn.compiler_mixing(should_mix, lang))
```
I would make this fact unary, call it something like:
```python
no_compiler_mixing(lang)
```
and emit it only if compiler mixing is disabled. Seeing a fact with a hard-coded `0` in the ASP is not very legible imo.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1570,6 +1570,35 @@ compiler_used_as_a_library(node(X, Child), Hash) :-
+% compiler within the root unification set "registers" that compiler
+% (and the root set is prohibited from having >= 2 compilers for that
+% language).
+unification_set_compiler("root", CompilerName, CompilerVersion, Language) :-
```
If I'm reading correctly, this allows using:
```
CompilerName@CompilerVersion +foo
CompilerName@CompilerVersion ~foo
```
as compilers when the `compiler_mixing` option is on. It would probably be a rare occurrence, but it's worth noting that somewhere if we keep the encoding like it is.

becker33 replied:
I don't see a good way to fix this without exposing build deps to the solver and probably having a huge performance impact -- I think we should document/avoid it for now.

scheibelp replied:
Originally I was identifying compilers by `node()` which seemed to work in a non-reuse context (1st 4 commits of this PR); for preventing mixing during reuse I was not able to identify compilers by `node()` or `hash`. In other words, what is generated now is something I can consistently generate across reused, external, and built packages which prevents mixing in all those cases (but does have an issue if two compilers that differ only by variants exist in the config).

becker33 replied:
We should probably document this as a known bug if we can't fix it for now.

scheibelp replied:
I added documentation on this at https://github.com/spack/spack/pull/51135/files#r2417500400

alalazo replied:
I think I found a way to fix this without exposing build dependencies, see `reused_provider` in https://github.com/spack/spack/pull/51417

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1570,6 +1570,35 @@ compiler_used_as_a_library(node(X, Child), Hash) :-
+    attr("version", node(CompilerID, CompilerName), CompilerVersion),
+    compiler_mixing(0, Language),
+    language(Language),
+    not allow_mixing(NodeName),
```
This needs a `#defined` to avoid:

<img width="1849" height="635" alt="Image" src="https://github.com/user-attachments/assets/8de20629-5462-4e15-9040-23297561f8c7" />

becker33 commented on lib/spack/spack/test/concretization/core.py:
```diff
@@ -468,6 +468,41 @@ def test_mixing_compilers_only_affects_subdag(self):
+        with spack.config.override("concretizer", {"compiler_mixing": False}):
+            with pytest.raises(spack.config.ConfigFormatError):
+                with spack.config.override("packages", {"all": {"allow_compiler_mixing": True}}):
+                    pass
```
```suggestion
                spack.config.set("packages", {"all": {"allow_compiler_mixing": True}})
```
No reason to create an extra code block for this

scheibelp committed:
simplify test

scheibelp committed:
avoid warning when fact is undefined

scheibelp committed:
move allow-mixing block to be next to global no-mixing block; rename fact and only register it as a negative

scheibelp committed:
rm redundant line in count (more than that, it was incorrect; but it was also unnecessary)

scheibelp committed:
update config: move everything into the concretizer schema (including exceptions by package name)

scheibelp committed:
add some documentation on this new feature

scheibelp committed:
add simple test for disabling compiler mixing with an env using unify:when_possible

scheibelp committed:
style fix

becker33 commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -280,3 +280,22 @@ Mixing Compilers
+   concretizer:
+     compiler_mixing: false
+
+This affects your root spec and everything it links to (transitively).
```
```suggestion
This affects root specs and any (transitive) link or run dependencies.
```

becker33 commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -280,3 +280,22 @@ Mixing Compilers
+This affects your root spec and everything it links to (transitively).
+Build-only dependencies are allowed to use different compilers (even when this is set).
+
+To enable mixing for specific packages, you can make that a whitelist:
```
```suggestion
Some packages are difficult to build with high performance compilers, and it may be necessary to enable compiler mixing just for those packages.
To enable mixing for specific packages, specify an allow-list in the ``compiler_mixing`` config:
```

becker33 commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -280,3 +280,22 @@ Mixing Compilers
+   concretizer:
+     compiler_mixing: ["openssl"]
+
+Note that this would not allow mixing for dependencies of ``openssl``.
```
```suggestion
.. note::
  Adding ``openssl`` to the compiler mixing allow-list does not allow mixing for dependencies of ``openssl``.
```

becker33 commented on lib/spack/spack/schema/concretizer.py:
```diff
@@ -61,6 +61,10 @@
             },
+            "compiler_mixing": {
+                "oneOf": [{"type": "boolean"}, {"type": "array"}],
+                "description": "Whether to allow compiler mixing between link dependencies",
```
```suggestion
                "description": "Whether to allow compiler mixing between link/run dependencies",
```

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3065,6 +3065,19 @@ def setup(
+                for pkg_name in should_mix:
+                    self.gen.fact(fn.allow_mixing(pkg_name))
+            except TypeError:
+                pass
```
```suggestion
        if should_mix is not True:  # anything besides should_mix: true
            for lang in ["c", "cxx", "fortran"]:
                self.gen.fact(fn.no_compiler_mixing(lang))

        if isinstance(should_mix, list):  # user specified an allow-list
            for pkg_name in should_mix:
                self.gen.fact(fn.allow_mixing(pkg_name))
```
I don't think there's anything logically wrong with the current implementation, but it's a bit hard to read. I'm pretty sure this is logically equivalent.

becker33 reviewed (changes requested):
Some comments on the new docs and one spot I think can be reworked for maintainability, but generally this LGTM at this point.

scheibelp committed:
review edits

scheibelp committed:
add note about limitations of mixing prevention

scheibelp commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -280,3 +280,30 @@ Mixing Compilers
+.. note::
+
+   Mixing detection is based on identifying compilers by name and version.
+   If you have two compilers that differ only in terms of variants, ``compiler_mixing:false`` would allow using both in a DAG.
```
I think this addresses https://github.com/spack/spack/pull/51135#discussion_r2417336472

scheibelp committed:
add test about linking to one compiler and compiling with another

scheibelp committed:
reorder clauses to match prior block where possible

scheibelp committed:
Possible test interaction is causing hang in CI (not locally) - I'm seeing if removing
this test leads to CI success.

Revert "add test about linking to one compiler and compiling with another"

This reverts commit 40dc56a66b4b7bf37a9e0a3209ec51af48da751e.

scheibelp committed:
Revert "Possible test interaction is causing hang in CI (not locally) - I'm seeing if removing"

This reverts commit 6e10ad57b39acabc188c1349003cbd2517067284.

scheibelp committed:
Merge branch 'develop' into disable-compiler-mixing2

scheibelp committed:
experiment with using 51417 to create a full compiler identity (not just name/version tuple)

scheibelp committed:
rm commented out bits

scheibelp committed:
dont need reused_provider to derive node() identity

scheibelp committed:
augment test to make sure that we can assign a matching compiler under a reuse scheme and it works; however the node ID in reuse scheme does not seem to sync with non-reuse

scheibelp committed:
all tests pass when using name/version tuples

scheibelp commented on lib/spack/spack/test/concretization/core.py:
```diff
@@ -468,6 +469,77 @@ def test_mixing_compilers_only_affects_subdag(self):
+                spack.concretize.concretize_one(f"dt-diamond%clang ^/{lefthash}")
+
+            # Should be able to reuse if the compilers match
+            spack.concretize.concretize_one(f"dt-diamond%gcc ^/{lefthash}")
```
This is the check that fails if using `node(CompilerHash, CompilerName)`, as mentioned in https://github.com/spack/spack/pull/51135#issuecomment-3413232347

scheibelp committed:
failure was from locking stage; mutable config implies mock stage; see if this bypasses issue

scheibelp committed:
mock store creation now also redirects stage via session-scoped fixture

scheibelp committed:
update comments, remove mutable_config fixture from individual test because it was applied from the containing class

scheibelp committed:
bypass mutable_database entirely and strip test_disable_mixing_reuse down to minimum infrastructure setup

scheibelp committed:
Revert other attempted fix, because 68aaa64 appears to have been sufficient.

Revert "bypass mutable_database entirely and strip test_disable_mixing_reuse down to minimum infrastructure setup"

This reverts commit ea9f71d25fce899c03f589af7e14a089b2fe1ca8.

scheibelp committed:
merge from develop (appears to be only minor sync issue)

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1688,6 +1688,37 @@ compiler_used_as_a_library(node(X, Child), Hash) :-
+    no_compiler_mixing(Language),
+    language(Language),
+    not allow_mixing(ParentName),
+    unification_set("root", ParentNode).
```
With reference to https://github.com/spack/spack/pull/51135#discussion_r2437745875, this should work:
```suggestion
unification_set_compiler("root", CompilerNode, Language) :-
    node_compiler(node(ID, Package), CompilerNode, Language),
    no_compiler_mixing(Language),
    not allow_mixing(Package),
    unification_set("root", node(ID, Package)).

compiler_hash(CompilerNode, Hash) :- node_compiler(_, CompilerNode, _), attr("hash", CompilerNode, Hash).

% Compiler for a reused node
reused_node_compiler(PackageNode, node(CompilerHash, Compiler), Language) :-
    concrete(PackageNode),
    attr("concrete_build_dependency", PackageNode, Compiler, CompilerHash),
    attr("virtual_on_build_edge", PackageNode, Compiler, Language).

% This compiler is not used by any "built" package, so use the Hash as the node ID
unification_set_compiler("root", node(CompilerHash, Compiler), Language) :-
    reused_node_compiler(node(ID, Package), node(CompilerHash, Compiler), Language),
    not compiler_hash(_, CompilerHash),
    no_compiler_mixing(Language),
    not allow_mixing(Package),
    unification_set("root", node(ID, Package)).

% This compiler is used by at least another built package, so use its node ID
unification_set_compiler("root", CompilerNode, Language) :-
    reused_node_compiler(node(ID, Package), node(CompilerHash, Compiler), Language),
    compiler_hash(CompilerNode, CompilerHash),
    no_compiler_mixing(Language),
    not allow_mixing(Package),
    unification_set("root", node(ID, Package)).
```

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3082,6 +3082,14 @@ def setup(
+                self.gen.fact(fn.no_compiler_mixing(lang))
+        if isinstance(should_mix, list):  # user specified an allow-list
+            for pkg_name in should_mix:
+                self.gen.fact(fn.allow_mixing(pkg_name))
```
(nitpick) If we extract a function, we can avoid the gotcha of the `is not` operator:
```diff
diff --git a/lib/spack/spack/solver/asp.py b/lib/spack/spack/solver/asp.py
index 6814447c2f..f178c5580d 100644
--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -3035,13 +3035,7 @@ def setup(
         # TODO: warning is because mypy doesn't know Spec supports rich comparison via decorator
         self.possible_compilers.sort()  # type: ignore[call-arg,call-overload]
 
-        should_mix = spack.config.get("concretizer:compiler_mixing", True)
-        if should_mix is not True:  # anything besides should_mix: true
-            for lang in ["c", "cxx", "fortran"]:
-                self.gen.fact(fn.no_compiler_mixing(lang))
-        if isinstance(should_mix, list):  # user specified an allow-list
-            for pkg_name in should_mix:
-                self.gen.fact(fn.allow_mixing(pkg_name))
+        self.compiler_mixing()
 
         self.gen.h1("Runtimes")
         injected_dependencies = self.define_runtime_constraints()
@@ -3171,6 +3165,18 @@ def setup(
 
         return self.gen
 
+    def compiler_mixing(self) -> None:
+        compiler_mixing = spack.config.get("concretizer:compiler_mixing", True)
+        if compiler_mixing is True:
+            return
+
+        for lang in ("c", "cxx", "fortran"):
+            self.gen.fact(fn.no_compiler_mixing(lang))
+
+        if isinstance(compiler_mixing, list):  # user specified an allow-list
+            for pkg_name in compiler_mixing:
+                self.gen.fact(fn.allow_mixing(pkg_name))
+
     def internal_errors(self, *, _use_unsat_cores: bool):
         parent_dir = os.path.dirname(__file__)

```

alalazo commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -281,3 +281,30 @@ Mixing Compilers
+.. note::
+
+   Mixing detection is based on identifying compilers by name and version.
+   If you have two compilers that differ only in terms of variants, ``compiler_mixing:false`` would allow using both in a DAG.
```
If we go with the implementation based on nodes, we can delete this:

```suggestion
```

scheibelp replied:
I tested node-based identity and it ran into https://github.com/spack/spack/pull/51135#discussion_r2437745875, so I think name/version-based identity will be needed.

EDIT: perhaps in https://github.com/spack/spack/pull/51135#discussion_r2445563338 you are suggesting a way to do this with node-identity - I'll check that.

scheibelp replied:
To confirm: this passes if I apply this and update the `count`. I'm set up a side meeting to resolve some questions I have about the logic, but something like this will land tomorrow in this PR and it should no longer rely on compiler name/version for identity.

alalazo commented on lib/spack/docs/configuring_compilers.rst:
```diff
@@ -281,3 +281,30 @@ Mixing Compilers
+   concretizer:
+     compiler_mixing: ["openssl"]
+
+.. note::
```
(nitpick, if you agree) We can probably put this outside of the note, just as a following sentence in the docs.

alalazo commented on lib/spack/spack/test/concretization/core.py:
```diff
@@ -468,6 +469,77 @@ def test_mixing_compilers_only_affects_subdag(self):
+    def test_disable_mixing_reuse(self, mutable_database):
+        # Install a spec
+        left = spack.concretize.concretize_one("dt-diamond-left %gcc")
+        PackageInstaller([left.package], fake=True, explicit=True).install()
```
We can avoid the `PackageInstaller` and run just the concretizer, if we use lower level API calls. An example of a test like that is e.g. `test_version_weight_and_provenance` that uses `solver.driver.solve` and thus doesn't need a store, or a db.

scheibelp replied:
I did that in https://github.com/spack/spack/pull/51135/commits/ea9f71d25fce899c03f589af7e14a089b2fe1ca8, and then reverted it because overall I'm inclined to say that bypassing the typical install interface makes the unit tests harder to read (people have to know that "everything works pretty much the same", and also such bypasses are susceptible to refactors where things implicitly become different, in a quiet manner that can render a unit test no-longer relevant). I can reinstate that commit, but doing so comes with all above caveats.

scheibelp replied:
Main thing is I'm curious if you find https://github.com/spack/spack/commit/ea9f71d25fce899c03f589af7e14a089b2fe1ca8 workable: I do not think I can use the same approach as `test_version_weight_and_provenance` (i.e. feeding in my pre-concretized specs as args to `reuse`) because the specparser depends on being able to actually query the specs from the DB (and IMO the test is more reliable when explicitly requesting the installed spec).

alalazo replied:
That's fine. On my machine it's faster than before, we can improve more later.

alalazo commented on lib/spack/spack/test/conftest.py:
```diff
@@ -500,6 +500,25 @@ def mock_stage(tmp_path_factory: pytest.TempPathFactory, monkeypatch, request):
         shutil.rmtree(new_stage, onerror=onerror)
 
 
+@pytest.fixture(scope="session")
+def mock_stage_for_database(tmp_path_factory: pytest.TempPathFactory, monkeypatch_session):
+    """A session-scoped analog of mock_stage, so that the mock_store
+    fixture uses its own stage vs. the global stage root for spack.
+    """
```
Is this related to this PR, or can we extract it as a bugfix in a separate PR?

scheibelp replied:
Without this change, this PR fails. No other PR I'm aware of fails. I think the way it works is to relieve contention that slows down all PR runs (but only happens to exceed thresholds in the case of this PR).

alalazo reviewed (commented):
I left a few comments. Did anybody ever worked out if there are edge cases with splicing that would not be covered?

I don't think we should deal with them here, in case there are, but it would be good to know if we have them.

scheibelp committed:
node-id based compiler identity replaces name/version identity, using suggestion from Max

add test that confirms need for third compiler mixing block in concretize.lp

checking to see what aspects of this new test are required

update comments; update docs now that we dont identify by name/version

auto style

alalazo commented:
> In short, I don't think so:

Sure, the main question for me was if this problem is meant to be fixed in this PR (I was a bit perplexed seeing that in the TODO list). I guess we are on the same page:
- This PR deals only with compiler mixing in the root unification set
- Other PRs will introduce the `%%` and `%%%` syntax to solve the other issue

scheibelp committed:
collect asp compiler mixing logic into separate method

scheibelp committed:
rm test_disable_mixing_reuse_and_built2 as it is actually not checking anything beyond test_disable_mixing_prevents_mixing

scheibelp committed:
reorganize rules (swap order of 2 and 3); more documentation

scheibelp committed:
doc change: integrate note into normal flow

scheibelp committed:
augment reused_node_compiler to ensure it is a compiler

scheibelp committed:
my new fake package depends on libelf, which changes the output of one of the tests for dependents

scheibelp committed:
Reapply ea9f71d (revert 7644707) - see if this speeds up the tests.

scheibelp committed:
use fake_db_install for other mixing test (remove all references to mutable_database introduced by this PR)

scheibelp committed:
my current best attempt at explaining these rules

alalazo reviewed (changes requested):
I think this PR is basically ready. The only polishing remaining is imo to:
- Update the PR title and description
- Add some default configuration for the new option to our YAML files

The last point in particular is needed so that users could know what is the default if they use:
```console
$ spack config [get|blame] concretizer
```
Right now that command shows no value for `concretize:compiler_mixing`.

scheibelp committed:
add default config option

scheibelp commented:
@alalazo I believe review requests are now addressed:

* `spack config blame concretizer` now shows default setting (`compiler_mixing: true`)
* Minor updates to PR description: i.e. second TODO is punted, and this applies to root unifications set
